File: frontend/src/game/questions.ts
import type { Question } from './types'

export const questions: Question[] = [
  { id: 'meaning', category: 'flow', prompt: 'What does SCM connect?', choices: [{ id: 'a', label: 'Products to customers' }, { id: 'b', label: 'Only the checkout' }], correctChoiceId: 'a', explanation: 'SCM helps products move from source to customer.', visual: 'FLOW' },
  { id: 'supplier', category: 'roles', prompt: 'Who provides parts?', choices: [{ id: 'a', label: 'A supplier' }, { id: 'b', label: 'A customer' }], correctChoiceId: 'a', explanation: 'Suppliers provide materials and parts.', visual: 'SUPPLY' },
  { id: 'warehouse', category: 'warehouse', prompt: 'Why use a warehouse?', choices: [{ id: 'a', label: 'Store products' }, { id: 'b', label: 'Make roads' }], correctChoiceId: 'a', explanation: 'Warehouses hold products safely until they are needed.', visual: 'STORE' },
  { id: 'transport', category: 'transport', prompt: 'Best for an urgent local delivery?', choices: [{ id: 'a', label: 'Bike courier' }, { id: 'b', label: 'Cargo ship' }], correctChoiceId: 'a', explanation: 'A bike courier is quick and flexible for short trips.', visual: 'MOVE' },
  { id: 'retailer', category: 'roles', prompt: 'What is a supermarket?', choices: [{ id: 'a', label: 'A retailer' }, { id: 'b', label: 'A mine' }], correctChoiceId: 'a', explanation: 'Retailers sell products to customers.', visual: 'SHOP' },
  { id: 'inventory', category: 'warehouse', prompt: 'Too many unsold coats are…', choices: [{ id: 'a', label: 'Excess stock' }, { id: 'b', label: 'A route' }], correctChoiceId: 'a', explanation: 'Too much stock uses space and money.', visual: 'COUNT' },
  { id: 'green', category: 'sustainability', prompt: 'What can cut delivery emissions?', choices: [{ id: 'a', label: 'Join nearby deliveries' }, { id: 'b', label: 'Send empty vans' }], correctChoiceId: 'a', explanation: 'Fewer journeys use less fuel.', visual: 'GREEN' },
  { id: 'robot', category: 'technology', prompt: 'What can a warehouse robot do?', choices: [{ id: 'a', label: 'Sort packages' }, { id: 'b', label: 'Write stories' }], correctChoiceId: 'a', explanation: 'Robots can move and sort items.', visual: 'ROBOT' },
  { id: 'customer', category: 'flow', prompt: 'Who uses the product?', choices: [{ id: 'a', label: 'The customer' }, { id: 'b', label: 'The loading dock' }], correctChoiceId: 'a', explanation: 'The customer receives and uses the product.', visual: 'USE' },
  { id: 'data', category: 'technology', prompt: 'Why use supply-chain data?', choices: [{ id: 'a', label: 'Make better plans' }, { id: 'b', label: 'Make boxes heavy' }], correctChoiceId: 'a', explanation: 'Data helps teams plan stock, routes, and timing.', visual: 'DATA' },
  { id: 'factory', category: 'roles', prompt: 'Where are products made?', choices: [{ id: 'a', label: 'A factory' }, { id: 'b', label: 'A checkout' }], correctChoiceId: 'a', explanation: 'Factories turn materials into finished products.', visual: 'MAKE' },
  { id: 'forecast', category: 'technology', prompt: 'What can a forecast help predict?', choices: [{ id: 'a', label: 'Customer demand' }, { id: 'b', label: 'The weather only' }], correctChoiceId: 'a', explanation: 'Demand forecasts help businesses prepare the right stock.', visual: 'PLAN' },
  { id: 'ship', category: 'transport', prompt: 'Which carries goods across oceans?', choices: [{ id: 'a', label: 'A delivery bike' }, { id: 'b', label: 'A cargo ship' }], correctChoiceId: 'b', explanation: 'Cargo ships move large loads across oceans.', visual: 'SHIP' },
  { id: 'reuse', category: 'sustainability', prompt: 'Which choice creates less waste?', choices: [{ id: 'a', label: 'Reuse boxes' }, { id: 'b', label: 'Throw out boxes' }], correctChoiceId: 'a', explanation: 'Reusing boxes keeps useful material in circulation.', visual: 'REUSE' },
  { id: 'quality', category: 'roles', prompt: 'What checks products work well?', choices: [{ id: 'a', label: 'Quality control' }, { id: 'b', label: 'A parking sign' }], correctChoiceId: 'a', explanation: 'Quality checks help catch problems before products ship.', visual: 'CHECK' },
  { id: 'lastmile', category: 'transport', prompt: 'What is the last mile?', choices: [{ id: 'a', label: 'Final trip to you' }, { id: 'b', label: 'A factory machine' }], correctChoiceId: 'a', explanation: 'The last mile is the final delivery to the customer.', visual: 'FINAL' },
  { id: 'barcode', category: 'technology', prompt: 'What can a barcode identify?', choices: [{ id: 'a', label: 'A product' }, { id: 'b', label: 'A road' }], correctChoiceId: 'a', explanation: 'Barcodes help teams identify and track products.', visual: 'SCAN' },
  { id: 'air', category: 'transport', prompt: 'Which is usually fastest for far-away urgent goods?', choices: [{ id: 'a', label: 'Air freight' }, { id: 'b', label: 'A slow boat' }], correctChoiceId: 'a', explanation: 'Air freight is fast, though it can use more energy.', visual: 'AIR' },
  { id: 'repair', category: 'sustainability', prompt: 'What helps products last longer?', choices: [{ id: 'a', label: 'Repair them' }, { id: 'b', label: 'Replace them fast' }], correctChoiceId: 'a', explanation: 'Repairing products can reduce waste and new materials.', visual: 'FIX' },
  { id: 'truck', category: 'transport', prompt: 'What moves goods between nearby cities?', choices: [{ id: 'a', label: 'A truck' }, { id: 'b', label: 'A bookshelf' }], correctChoiceId: 'a', explanation: 'Trucks are flexible for road deliveries.', visual: 'TRUCK' },
  { id: 'safety', category: 'warehouse', prompt: 'What keeps a warehouse safe?', choices: [{ id: 'a', label: 'Clear walkways' }, { id: 'b', label: 'Boxes in every path' }], correctChoiceId: 'a', explanation: 'Clear walkways help people and equipment move safely.', visual: 'SAFE' },
]

export function validateQuestions(items: Question[]): void {
  const ids = new Set<string>()
  items.forEach((question) => {
    if (ids.has(question.id)) throw new Error(`Duplicate question id: ${question.id}`)
    ids.add(question.id)
    if (question.choices.length !== 2) throw new Error(`Question ${question.id} must have exactly two choices`)
    if (!question.choices.some((choice) => choice.id === question.correctChoiceId)) throw new Error(`Question ${question.id} has an invalid correct choice`)
  })
}

validateQuestions(questions)

File: python_app/questions.py
"""Question bank and validation for Supply Chain Quest."""

from dataclasses import dataclass
from typing import Sequence


@dataclass(frozen=True)
class AnswerChoice:
    id: str
    label: str


@dataclass(frozen=True)
class Question:
    id: str
    category: str
    prompt: str
    choices: tuple[AnswerChoice, AnswerChoice]
    correct_choice_id: str
    explanation: str
    visual: str


questions: tuple[Question, ...] = (
    Question("meaning", "flow", "What does SCM connect?", (AnswerChoice("a", "Products to customers"), AnswerChoice("b", "Only the checkout")), "a", "SCM helps products move from source to customer.", "FLOW"),
    Question("supplier", "roles", "Who provides parts?", (AnswerChoice("a", "A supplier"), AnswerChoice("b", "A customer")), "a", "Suppliers provide materials and parts.", "SUPPLY"),
    Question("warehouse", "warehouse", "Why use a warehouse?", (AnswerChoice("a", "Store products"), AnswerChoice("b", "Make roads")), "a", "Warehouses hold products safely until they are needed.", "STORE"),
    Question("transport", "transport", "Best for an urgent local delivery?", (AnswerChoice("a", "Bike courier"), AnswerChoice("b", "Cargo ship")), "a", "A bike courier is quick and flexible for short trips.", "MOVE"),
    Question("retailer", "roles", "What is a supermarket?", (AnswerChoice("a", "A retailer"), AnswerChoice("b", "A mine")), "a", "Retailers sell products to customers.", "SHOP"),
    Question("inventory", "warehouse", "Too many unsold coats are...", (AnswerChoice("a", "Excess stock"), AnswerChoice("b", "A route")), "a", "Too much stock uses space and money.", "COUNT"),
    Question("green", "sustainability", "What can cut delivery emissions?", (AnswerChoice("a", "Join nearby deliveries"), AnswerChoice("b", "Send empty vans")), "a", "Fewer journeys use less fuel.", "GREEN"),
    Question("robot", "technology", "What can a warehouse robot do?", (AnswerChoice("a", "Sort packages"), AnswerChoice("b", "Write stories")), "a", "Robots can move and sort items.", "ROBOT"),
    Question("customer", "flow", "Who uses the product?", (AnswerChoice("a", "The customer"), AnswerChoice("b", "The loading dock")), "a", "The customer receives and uses the product.", "USE"),
    Question("data", "technology", "Why use supply-chain data?", (AnswerChoice("a", "Make better plans"), AnswerChoice("b", "Make boxes heavy")), "a", "Data helps teams plan stock, routes, and timing.", "DATA"),
    Question("factory", "roles", "Where are products made?", (AnswerChoice("a", "A factory"), AnswerChoice("b", "A checkout")), "a", "Factories turn materials into finished products.", "MAKE"),
    Question("forecast", "technology", "What can a forecast help predict?", (AnswerChoice("a", "Customer demand"), AnswerChoice("b", "The weather only")), "a", "Demand forecasts help businesses prepare the right stock.", "PLAN"),
    Question("ship", "transport", "Which carries goods across oceans?", (AnswerChoice("a", "A delivery bike"), AnswerChoice("b", "A cargo ship")), "b", "Cargo ships move large loads across oceans.", "SHIP"),
    Question("reuse", "sustainability", "Which choice creates less waste?", (AnswerChoice("a", "Reuse boxes"), AnswerChoice("b", "Throw out boxes")), "a", "Reusing boxes keeps useful material in circulation.", "REUSE"),
    Question("quality", "roles", "What checks products work well?", (AnswerChoice("a", "Quality control"), AnswerChoice("b", "A parking sign")), "a", "Quality checks help catch problems before products ship.", "CHECK"),
    Question("lastmile", "transport", "What is the last mile?", (AnswerChoice("a", "Final trip to you"), AnswerChoice("b", "A factory machine")), "a", "The last mile is the final delivery to the customer.", "FINAL"),
    Question("barcode", "technology", "What can a barcode identify?", (AnswerChoice("a", "A product"), AnswerChoice("b", "A road")), "a", "Barcodes help teams identify and track products.", "SCAN"),
    Question("air", "transport", "Which is usually fastest for far-away urgent goods?", (AnswerChoice("a", "Air freight"), AnswerChoice("b", "A slow boat")), "a", "Air freight is fast, though it can use more energy.", "AIR"),
    Question("repair", "sustainability", "What helps products last longer?", (AnswerChoice("a", "Repair them"), AnswerChoice("b", "Replace them fast")), "a", "Repairing products can reduce waste and new materials.", "FIX"),
    Question("truck", "transport", "What moves goods between nearby cities?", (AnswerChoice("a", "A truck"), AnswerChoice("b", "A bookshelf")), "a", "Trucks are flexible for road deliveries.", "TRUCK"),
    Question("safety", "warehouse", "What keeps a warehouse safe?", (AnswerChoice("a", "Clear walkways"), AnswerChoice("b", "Boxes in every path")), "a", "Clear walkways help people and equipment move safely.", "SAFE"),
)


def validate_questions(items: Sequence[Question]) -> None:
    seen_ids: set[str] = set()
    for question in items:
        if question.id in seen_ids:
            raise ValueError(f"Duplicate question id: {question.id}")
        seen_ids.add(question.id)
        if len(question.choices) != 2:
            raise ValueError(f"Question {question.id} must have exactly two choices")
        choice_ids = {choice.id for choice in question.choices}
        if question.correct_choice_id not in choice_ids:
            raise ValueError(f"Question {question.id} has an invalid correct choice")


validate_questions(questions)
